File: server/src/au/com/codeka/warworlds/server/Configuration.java
package au.com.codeka.warworlds.server;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.Expose;
import com.google.gson.stream.JsonReader;

/**
 * The server's configuration parameters are read from a json object on startup and parsed into
 * this object via Gson.
 */
public class Configuration {
  public static Configuration i;

  public static final String PROD_CLIENT_ID =
      "1021675369049-85dehn126ib087kkc270k0lko6ahv2h7.apps.googleusercontent.com";
  public static final String DEV_CLIENT_ID =
      "1021675369049-kh3j8g9m8ugkrqamllddh3v0coss7gc8.apps.googleusercontent.com";

  /** Loads the {@link Configuration} from the given file and sets it to {@code Configuration.i}. */
  public static void loadConfig() throws FileNotFoundException {
    String fileName = System.getProperty("au.com.codeka.warworlds.server.ConfigFile");
    if (fileName == null) {
      // just try whatever in the current directory
      fileName = "config.json";
    }

    Gson gson = new GsonBuilder().create();
    JsonReader jsonReader = new JsonReader(new FileReader(fileName));
    jsonReader.setLenient(true); // allow comments (and a few other things)
    i = gson.fromJson(jsonReader, Configuration.class);
  }

  @Expose private String realmName;
  @Expose private String dataDirectory;
  @Expose private String configDirectory;
  @Expose private int listenPort;
  @Expose private Integer numStarSimulationThreads;
  @Expose private DatabaseConfiguration database;
  @Expose private LimitsConfiguration limits;
  @Expose private ResetsConfig resets;
  @Expose private PatreonConfig patreon;
  @Expose private String requestStatsDirectory;
  @Expose private boolean allowNonProdClientLogins;
  @Expose private ClickerConfig[] clickers;
  @Expose private SafetyNetConfig safetyNet;

  public Configuration() {
    limits = new LimitsConfiguration();
  }

  public String getRealmName() {
    return realmName;
  }

  public File getDataDirectory() {
    return new File(dataDirectory);
  }

  public File getConfigDirectory() { return new File(configDirectory); }

  public int getListenPort() {
    return listenPort;
  }

  public int getNumStarSimulationThreads() {
    if (numStarSimulationThreads == null) {
      return 1;
    }
    return numStarSimulationThreads.intValue();
  }

  public DatabaseConfiguration getDatabaseConfig() {
    return database;
  }

  public ResetsConfig getResets() {
    return resets;
  }

  public LimitsConfiguration getLimits() {
    return limits;
  }

  public PatreonConfig getPatreon() {
    return patreon;
  }

  public String getRequestStatsDirectory() {
    return requestStatsDirectory;
  }

  /** If true, we'll allow non-prod clients to log in. Default is false. */
  public boolean getAllowNonProdClientLogins() {
    return allowNonProdClientLogins;
  }

  public ClickerConfig[] getClickers() {
    return clickers;
  }

  public SafetyNetConfig getSafetyNet() {
    return safetyNet;
  }

  public static class DatabaseConfiguration {
    @Expose private String server;
    @Expose private int port;
    @Expose private String database;
    @Expose private String schema;
    @Expose private String username;
    @Expose private String password;

    public String getServer() {
      return server;
    }

    public int getPort() {
      return port;
    }

    public String getDatabase() {
      return database;
    }

    public String getSchema() {
      return schema;
    }

    public String getUsername() {
      return username;
    }

    public String getPassword() {
      return password;
    }
  }

  public static class LimitsConfiguration {
    @Expose private int maxEmpireNameLength;
    @Expose private int maxAllianceNameLength;
    @Expose private int maxStarNameLength;
    @Expose private double maxEmojiRatio;

    public LimitsConfiguration() {
      maxEmpireNameLength = 40;
      maxAllianceNameLength = 80;
      maxStarNameLength = 40;
      maxEmojiRatio = 0.2;
    }

    public int maxEmpireNameLength() {
      return maxEmpireNameLength;
    }

    public int maxAllianceNameLength() {
      return maxAllianceNameLength;
    }

    public int maxStarNameLength() {
      return maxStarNameLength;
    }

    public double getMaxEmojiRatio() {
      return maxEmojiRatio;
    }
  }

  public static class ResetsConfig {
    @Expose private int resetPeriodDays;
    @Expose private int maxResetsPerPeriod;

    public int getResetPeriodDays() {
      return resetPeriodDays;
    }

    public int getMaxResetsPerPeriod() {
      return maxResetsPerPeriod;
    }
  }

  public static class PatreonConfig {
    @Expose private String clientId;
    @Expose private String clientSecret;
    @Expose private String redirectUri;
    @Expose private String refreshToken;

    public String getClientId() {
      return clientId;
    }

    public String getClientSecret() {
      return clientSecret;
    }

    public String getRedirectUri() {
      return redirectUri;
    }

    public String getRefreshToken() {
      return refreshToken;
    }
  }

  public static class ClickerConfig {
    @Expose private String name;
    @Expose private String appName;

    public String getName() {
      return name;
    }

    public String getAppName() {
      return appName;
    }
  }

  public static class SafetyNetConfig {
    @Expose boolean enabled;
    @Expose int[] exemptedEmpires;

    public boolean isEnabled() {
      return enabled;
    }

    public int[] getExemptedEmpires() {
      return exemptedEmpires;
    }
  }
}

File: server/src/main/java/au/com/codeka/warworlds/server/Configuration.kt
package au.com.codeka.warworlds.server

import au.com.codeka.warworlds.common.Log
import com.google.auth.oauth2.GoogleCredentials
import com.google.common.collect.Lists
import com.google.gson.GsonBuilder
import com.google.gson.InstanceCreator
import com.google.gson.JsonElement
import com.google.gson.annotations.Expose
import com.google.gson.stream.JsonReader
import java.io.ByteArrayInputStream
import java.io.FileReader
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.lang.reflect.Type
import java.nio.charset.StandardCharsets

/**
 * The server's configuration parameters are read from a json object on startup and parsed into
 * this object via Gson.
 */
class Configuration private constructor() {
  @Expose
  val baseUrl: String? = null

  @Expose
  val listenPort = 0

  @Expose
  val smtp: SmtpConfig

  @Expose
  private val firebase: JsonElement? = null

  @Expose
  val patreon: PatreonConfig? = null

  @Expose
  val limits: LimitsConfig? = null
  private var firebaseCredentials: GoogleCredentials? = null

  /** Loads the [Configuration] from the given file and sets it to `Configuration.i`.  */
  @Throws(IOException::class)
  fun load() {
    var fileName = System.getProperty("ConfigFile")
    if (fileName == null) {
      // just try whatever in the current directory
      fileName = "config.json"
    }
    log.info("Loading config from: %s", fileName)
    val gson = GsonBuilder()
        .registerTypeAdapter(Configuration::class.java, InstanceCreator { type: Type? -> i } as InstanceCreator<Configuration>)
        .create()
    val jsonReader = JsonReader(FileReader(fileName))
    jsonReader.isLenient = true // allow comments (and a few other things)
    gson.fromJson<Any>(jsonReader, Configuration::class.java)
  }

  fun getFirebaseCredentials(): GoogleCredentials? {
    try {
      if (firebaseCredentials == null) {
        try {
          firebaseCredentials = GoogleCredentials.fromStream(
              ByteArrayInputStream(firebase.toString().toByteArray(StandardCharsets.UTF_8)))
              .createScoped(FIREBASE_SCOPES)
        } catch (e: UnsupportedEncodingException) {
          // Should never happen.
        }
      }
      firebaseCredentials!!.refreshIfExpired()
    } catch (e: IOException) {
      throw RuntimeException("Should never happen.", e)
    }
    return firebaseCredentials
  }

  class SmtpConfig {
    @Expose
    val host = "smtp-relay.gmail.com"

    @Expose
    val port = 587

    @Expose
    val userName: String? = null

    @Expose
    val password: String? = null

    @Expose
    val senderAddr = "noreply@war-worlds.com"

  }

  class PatreonConfig {
    @Expose
    val clientId: String? = null

    @Expose
    val clientSecret: String? = null

    @Expose
    val redirectUri: String? = null

  }

  class LimitsConfig {
    @Expose
    val maxEmpireNameLength = 0
  }

  companion object {
    val i = Configuration()
    private val log = Log("Configuration")
    private val FIREBASE_SCOPES: Collection<String> = Lists.newArrayList("https://www.googleapis.com/auth/firebase.messaging")
  }

  init {
    smtp = SmtpConfig()
  }
}